library/alloc/src/collections/btree/map/tests.rs

```diff
@@ -1822,6 +1822,31 @@ fn test_vacant_entry_key() {
     a.check();
 }
 
+#[test]
+fn test_vacant_entry_no_insert() {
+    let mut a = BTreeMap::<&str, ()>::new();
+    let key = "hello there";
+
+    // Non-allocated
+    assert!(a.root.is_none());
+    match a.entry(key) {
+        Occupied(_) => unreachable!(),
+        Vacant(e) => assert_eq!(key, *e.key()),
+    }
+    assert!(a.root.is_none());
+    a.check();
+
+    // Allocated
+    BTreeMap::ensure_has_root(&mut a.root);
```

Comment:
I never properly realised you can call modifying private methodsin unit tests. I mean, it's obvious the compiler allows it, but I don't like it. It seems unclear when the test applies, how you could end up in this state using the public API (i,e, insert a key and remove it again).And there's more coupling. But since these is no way to stop it, is that a valid argument?

Reply:
Yes, we cannot stop people abusing `ensure_is_owned` like what I did. How about removing it and inline it everywhere?

For `Clone::clone`, now that the `out_tree` must be allocated, so I prefer slightly to use `unwrap` or use `if let Some(out_root) = out_tree.root` to guard that `out_tree.root` must be `Some`.

Including test codes, there is one more usage of `ensure_is_owned`. I think it can also be removed by testing if `iter.is_empty()` to avoid unnecessary reallocation.
https://github.com/rust-lang/rust/blob/86067bb461d044ee30e7880ab6f3b34d5070f1db/library/alloc/src/collections/btree/map/tests.rs#L118

Then remain `BTreeMap::append` and `Recover::replace`. `ensure_is_owned` transforms the root node from maybe-allocated to allocated. It is exactly where `ensure_is_owned` is supposed to be used. I think it may not be too verbose to inline them into `map.root.get_or_insert_with(Root::new)`, is it? Or alternatively, renaming with `get_or_allocate(&mut map.root)` might make better sense?

Reply:
> How about removing it and inline it everywhere?

Indeed

> For `Clone::clone`, now that the `out_tree` must be allocated, so I prefer slightly to use `unwrap`

Definitely. The only doubt to me is whether to use `unwrap` or `unwrap_unchecked`, and I vote `unwrap` because it's a tiny test compared to multiple allocations.

> Including test codes, there is one more usage of `ensure_is_owned`. I think it can also be removed by testing if `iter.is_empty()` to avoid unnecessary reallocation.

Good point. Incidentally, this is a function that I think belongs in the public API, for those who build maps with a long and stable life. All the more since #88448. You could start that by moving it as a private member of `BTreeMap`.

> it may not be too verbose to inline them into `map.root.get_or_insert_with(Root::new)`, is it?

Not at all.

> Or alternatively, renaming with `get_or_allocate(&mut map.root)` might make better sense?

Yes, better than the "ensure" verb.

Reply:
I have inlined `ensure_is_owned` everywhere, and I found that the inlined version `map.root.get_or_insert_with(Root::new)` is even shorter than `BTreeMap::ensure_is_owned(&mut map.root)`.

I also added two helper functions `has_root` and `has_no_root` to substitude the `root.is_some()` and `root.is_none()` in test codes.

> You could start that by moving it as a private member of `BTreeMap`.

I cannot move `compact` as a private member of `BTreeMap` without making it `pub` because it is not used by other functions.

```
error: associated function is never used: `compact`
    --> library/alloc/src/collections/btree/map.rs:2260:8
     |
2260 |     fn compact(&mut self)
     |        ^^^^^^^
     |
     = note: `-D dead-code` implied by `-D warnings`
```

Reply:
> I also added two helper functions `has_root`

Good idea. but now I remember I added `height` with the same concern in mind.

> > You could start that by moving it as a private member of `BTreeMap`.
> 
> I cannot move `compact` as a private member of `BTreeMap` without making it `pub` because it is not used by other functions.

Sigh… so now we know why I put it where it is.

Reply:
> > I also added two helper functions has_root
>
> Good idea. but now I remember I added height with the same concern in mind.

Oh, I seemed to forget applying that patch, I check the `height` method later.

Reply:
By the way, I'm not saying `height()` is better, I just remembered it. It happened to combine two related properties in a way I found nice enough. We get:

```assert_eq!(map.height(), None)```

which looks a bit odd at first, but when you learn that `node` defines height starting from 0 and the fact there may not be a root at all, it's a line that tests two properties for the price of one.

In your tests, that don't play with tree height, it just looks odd and `has_root` was more clear. `has_no_root` on the other hand, I find weird. Granted, ```assert!(!map.has_root())``` looks ugly, but we could write ```assert!(map.has_root().not())``` if we `use std::ops::Not`.

But I'm not saying you should change it back. In fact, I don't mind non-modifying calls like `map.root.is_none()` as long as 
there aren't too many of them.

Reply:
I added a comment to say what the `a.height()` assertion is test for. For the emptied allocated root, we can be more confident that the root remains unchanged, because it's height is also unchanged, besides the length of the tree (by the `is_empty()` test).

Reply:
> Granted, `assert!(!map.has_root())` looks ugly, but we could write `assert!(map.has_root().not())` if we use `std::ops::Not`.

I really love this style. I haven't come up with a good idea what to deal with the two almost consecutive `!`s, until you mentioned `std::ops::Not`, what a good taste! Thank you :D